File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        count = 0

        def dfs(node, curr_max):
            nonlocal count

            if not node:
                return 
            
            if node.val >= curr_max:
                count += 1
            new_max = max(curr_max, node.val)

            dfs(node.left, new_max)
            dfs(node.right, new_max)

        
        dfs(root, root.val)
        return count

File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        // base case 1: if node left and right is nullptr then its a leaf node. 
        if (!root) {
            return nullptr; 
        }

        //base case 2: if 
        if (root->left == nullptr and root->right == nullptr) {
            return root; 
        } else {
            TreeNode* temp = root->left;
            root->left = root->right;
            root->right = temp; 
        }
            invertTree(root->right);
            invertTree(root->left); 
        return root; 
  
    }
};
